designdocs/CAPACITY.md

```diff
@@ -0,0 +1,404 @@
+    
+    Ok(ValidTransaction { priority, ..Default::default() })
+  }
+
+  /// Check and Withdraw fee payment from Capacity or Token account before executing authoring.
+  fn pre_dispatch(
+    self,
+    who: &Self::AccountId,
+    call: &Self::Call,
+    info: &DispatchInfoOf<Self::Call>,
+    len: usize,
+  ) -> Result<Self::Pre, TransactionValidityError> {
+    let (_fee, initial_payment) = self.withdraw_fee(who, call, info, len)?;
+
+    let tip = self.0;
+
+    Ok((tip, who.clone(), initial_payment))
+  }
+
+  ...
+}
+```
+
+Notice that Pre-dispatch returns a type `Pre`, this is the type that gets passed from Pre-dispatch to Post-dispatch. It lets Post-dispach know how the payment was paid for in Capacity, Token or No payment.
+
+```rust
+/// Types for handling how the payment will processed. This type is passed to Post-dispatch to be able to distinguish how to reinbused overcharges in fee payment.
+#[derive(Encode, Decode, DefaultNoBound, TypeInfo)]
+pub enum InitialPayment<T: Config> {
+  /// Pay no fee.
+  Nothing,
+  /// Pay fee with Token.
+  Token(LiquidityInfoOf<T>),
+  /// Pay fee with Capacity.
+  Capacity(ChargeCapacityBalanceOf<T>),
+}
+```
+
+When after the transaction is authored Post-dispatch refunds the overcharged Capacity or Token payment. Using the type `Pre` that gets passed in from Pre-dispatch it corrects the fee refunding the amount overcharged.
+
+```rust
+impl<T: Config + Send + Sync> SignedExtension for ChargeTransactionPayment<T>
+where
+	BalanceOf<T>: Send + Sync + FixedPointOperand + From<u64> + IsType<ChargeCapacityBalanceOf<T>>,
+	CallOf<T>: Dispatchable<Info = DispatchInfo, PostInfo = PostDispatchInfo> + IsSubType<Call<T>>,
+{
+  ...
+  
+  fn post_dispatch(
+    pre: Self::Pre,
+    info: &DispatchInfoOf<Self::Call>,
+    post_info: &PostDispatchInfoOf<CallOf<T>>,
+    len: usize,
+    result: &DispatchResult,
+  ) -> Result<(), TransactionValidityError> {
+    use pallet_transaction_payment::ChargeTransactionPayment;
+    
+    let (tip, who, initial_payment) = pre;
+    let actual_fee = pallet_transaction_payment::Pallet::<T>::compute_actual_fee(
+      len as u32,
+      info,
+      post_info,
+      tip,
+    );
+    
+    match initial_payment {
+      InitialPayment::Token(already_withdrawn) => {
+        let pre_pre = (tip, who, already_withdrawn);
+        ChargeTransactionPayment::<T>::post_dispatch(
+          pre_pre, info, post_info, len, result,
+        )?;
+      },
+      InitialPayment::Capacity(already_withdrawn) => {
+        let account = 1u32; 
+        T::OnChargeCapacityTransaction::correct_and_deposit_fee(
+          &account,
+          info,
+          post_info,
+          actual_fee.into(),
+          tip.into(),
+          already_withdrawn,
+        )?;
+      },
+      InitialPayment::Nothing => {
+        debug_assert!(tip.is_zero(), "tip should be zero if initial fee was zero.");
+      },
+    }
+    Ok(())
+  }
+  ...
+}
+```
```

Comment:
Agree we dont need this in design doc, implementation should depend of story holder and put their skills to practice